File: backend/app/schemas/consolidation.py
"""
Pydantic schemas for the Admin Consolidated PPMP view.

Pure READ/response models — nothing here is persisted.

Matches the REAL app.models.ppmp shape (confirmed via mongosh against
live data):
- PPMP.office_id stores a FeeCategoryOffice._id directly (see
  consolidation_service.py for the full explanation).
- PPMPProject has no title field — only order_no, remarks,
  attached_document_title, supporting_docs, total_budget. Projects are
  labeled using order_no (or remarks, if present).
- PPMPEntry carries the full procurement detail set (project_type,
  procurement_mode, timeline fields, source_of_funds, etc.) — these are
  now passed through so the consolidated view can render the same table
  layout as PPMPDetailPage.tsx, not just a flattened item list.
- Each PPMP has its own description/additional_description/signatories.
  When an office has more than one matching PPMP for a given
  year/type, the FIRST one's meta fields (description, signatories,
  ppmp_no) are used to represent the office — consistent with the
  existing convention already used for ppmp_id/ppmp_no below.
"""
from __future__ import annotations

from typing import Optional
from pydantic import BaseModel, Field


class ConsolidationFilters(BaseModel):
    fee_category: str = Field(..., description="e.g. 'OJT Fees' — matched via FeeCategoryOffice, not entry text")
    fiscal_year: int = Field(..., description="Matched against PPMP.year")
    ppmp_type: str = Field(..., description="'indicative' or 'final'")
    search: Optional[str] = None


class ConsolidatedItemOut(BaseModel):
    item_name: str
    quantity: float
    unit: str
    unit_price: float
    total_cost: float


class ConsolidatedEntryOut(BaseModel):
    entry_id: str
    category_description: str        # e.g. "Spare Parts" — entry-level code, unrelated to Fee Category
    description: str                 # the entry's own specific description
    project_type: str = ""
    procurement_mode: str = ""
    pre_proc_conference: str = ""
    start_activity: str = ""
    end_activity: str = ""
    delivery_period: str = ""
    source_of_funds: str = ""
    items: list[ConsolidatedItemOut]
    entry_subtotal: float


class ConsolidatedProjectOut(BaseModel):
    project_id: str                   # synthesized as f"{ppmp_id}-{order_no}" — PPMPProject has no id of its own
    project_label: str                # e.g. "Project 1", or project.remarks if present
    remarks: Optional[str] = None
    attached_document_title: str = ""
    entries: list[ConsolidatedEntryOut]
    project_subtotal: float


class ConsolidatedSignatoryOut(BaseModel):
    sign_off: str
    name: str
    position: str
    order_no: int


class ConsolidatedOfficeOut(BaseModel):
    office_id: str
    office_name: str
    ppmp_id: str
    ppmp_no: Optional[str] = None
    ppmp_type: str
    fiscal_year: int
    description: str = ""
    additional_description: str = ""
    signatories: list[ConsolidatedSignatoryOut] = []
    projects: list[ConsolidatedProjectOut]
    office_total: float


class ConsolidatedPPMPResponse(BaseModel):
    fee_category: str
    fiscal_year: int
    ppmp_type: str
    offices: list[ConsolidatedOfficeOut]
    grand_total: float
    office_count: int
    generated_at: str

File: frontend/src/services/consolidation.ts
import api from "./api";

export interface ConsolidatedItem {
  item_name: string;
  quantity: number;
  unit: string;
  unit_price: number;
  total_cost: number;
}

export interface ConsolidatedEntry {
  entry_id: string;
  category_description: string; // entry-level code, e.g. "Spare Parts" — unrelated to Fee Category
  description: string; // the entry's own specific description
  project_type: string;
  procurement_mode: string;
  pre_proc_conference: string;
  start_activity: string;
  end_activity: string;
  delivery_period: string;
  source_of_funds: string;
  items: ConsolidatedItem[];
  entry_subtotal: number;
}

export interface ConsolidatedProject {
  project_id: string; // synthesized as `${ppmp_id}-${order_no}` — PPMPProject has no id of its own
  project_label: string; // "Project N", or the project's remarks if set — PPMPProject has no title field
  remarks?: string | null;
  attached_document_title: string;
  entries: ConsolidatedEntry[];
  project_subtotal: number;
}

export interface ConsolidatedSignatory {
  sign_off: string;
  name: string;
  position: string;
  order_no: number;
}

export interface ConsolidatedOffice {
  office_id: string;
  office_name: string;
  ppmp_id: string;
  ppmp_no?: string;
  ppmp_type: string;
  fiscal_year: number;
  description: string;
  additional_description: string;
  signatories: ConsolidatedSignatory[];
  projects: ConsolidatedProject[];
  office_total: number;
}

export interface ConsolidatedPPMPResponse {
  fee_category: string;
  fiscal_year: number;
  ppmp_type: string;
  offices: ConsolidatedOffice[];
  grand_total: number;
  office_count: number;
  generated_at: string;
}

// ── API calls ───────────────────────────────────────────────────────────

export interface ConsolidationFilters {
  feeCategory: string;
  fiscalYear: number;
  ppmpType: string;
}

// Returns the real Fee Category list (same names shown in the Fee
// Categories admin tab — STF, OJT Fees, Laboratory Fees, etc.), not
// anything derived from PPMP entry text.
export async function fetchFeeCategories(
  requesterUid: string,
): Promise<string[]> {
  const { data } = await api.get("/admin/ppmp-consolidation/categories", {
    params: { requester_uid: requesterUid },
  });
  return data;
}

export async function fetchConsolidatedPPMP(
  filters: ConsolidationFilters,
  requesterUid: string,
): Promise<ConsolidatedPPMPResponse> {
  const { data } = await api.get("/admin/ppmp-consolidation", {
    params: {
      requester_uid: requesterUid,
      fee_category: filters.feeCategory,
      fiscal_year: filters.fiscalYear,
      ppmp_type: filters.ppmpType,
    },
  });
  return data;
}

function buildExportUrl(
  kind: "excel" | "pdf",
  filters: ConsolidationFilters,
  requesterUid: string,
): string {
  const params = new URLSearchParams({
    requester_uid: requesterUid,
    fee_category: filters.feeCategory,
    fiscal_year: String(filters.fiscalYear),
    ppmp_type: filters.ppmpType,
  });
  return `/admin/ppmp-consolidation/export/${kind}?${params.toString()}`;
}

export async function downloadConsolidatedExport(
  kind: "excel" | "pdf",
  filters: ConsolidationFilters,
  requesterUid: string,
): Promise<void> {
  const response = await api.get(buildExportUrl(kind, filters, requesterUid), {
    responseType: "blob",
  });
  const blob = new Blob([response.data]);
  const url = window.URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  const ext = kind === "excel" ? "xlsx" : "pdf";
  a.download =
    `PPMP_Consolidation_${filters.feeCategory}_${filters.fiscalYear}_${filters.ppmpType}.${ext}`.replace(
      /\s+/g,
      "_",
    );
  document.body.appendChild(a);
  a.click();
  a.remove();
  window.URL.revokeObjectURL(url);
}
